tensorboard/plugins/profile/pod_viewer/topology_graph/topology-graph.ts

```diff
@@ -0,0 +1,544 @@
+                  .style('opacity', 1.0);
+              d3.select('#tooltip').classed('hidden', true);
+            })
+        .transition()
+        .duration(1000)
+        .style('fill', (d) => colorScale(d.value / d.total));
+    cards.exit().remove();
+  },
+  /**
+   * Draw the UI of chip to chip links.
+   */
+  drawLinks: function(linkData) {
+    let parent = this;
+    if (!linkData || linkData.length == 0 || !this.gLink_) {
+      return;
+    }
+
+    // Handle links;
+    let links = this.gLink_.selectAll('.link').data(linkData);
+
+    // attach the arrow from defs
+    links.enter()
+        .append('svg:path')
+        .attr('id', (d) => 'cid' + d.channelId)
+        .attr('stroke-width', 2)
+        .attr('stroke', 'red')
+        .attr('fill', 'none')
+        .attr('marker-end', 'url(#arrow)')
+        .style('visibility', 'hidden')
+        .merge(links)
+        .attr('d', (d) => parent.linkToPath(d));
+
+    // Handle deleted links.
+    links.exit().remove();
+    this.selectedChannelIdChanged_(this.selectedChannelId);
+  },
+  /**
+   * Given the global core id, returns the (x, y) coordinates in the topology
+   * graph.
+   * @return [x, y]
+   */
+  coreIdToPos: function(id: number): number[] {
+    let p = this;
+    const chipId = Math.floor(id / 2);
+    const nodeId = id & 1;
+    const xDim = chipId % p.xDimension_;
+    const yDim = Math.floor(chipId / p.xDimension_);
+    const x =
+        p.getNodeXLoc(
+            Math.floor(xDim / p.hostXStride_), xDim % p.hostXStride_, nodeId) +
+        p.chipGridSize_ / p.nodesPerChip_ / 2;
+    const y = p.getChipYLoc(
+                  Math.floor(yDim / p.hostYStride_), yDim % p.hostYStride_) +
+        p.chipGridSize_ / 2;
+    return [x, y];
+  },
+  /**
+   * Returns the svg path given the src and dst core and node id.
+   * @return Path in svg format.
+   */
+  linkToPath: function(link): string {
+    let p = this;
+    const src = p.coreIdToPos(link.srcCoreId);
+    const dst = p.coreIdToPos(link.dstCoreId);
+    const path = 'M ' + src[0] + ' ' + src[1] + 'L ' + dst[0] + ' ' + dst[1];
+    return path;
+  },
+  /**
+   * Returns the text to visualize in the tool tips.
+   * @return String to render in tool tips.
+   */
+  getToolTipText_: function(data): string {
+    let parent = this;
+    let res = 'pos: (' + data.ydim + ',' + data.xdim + ')\n';
+    res += 'host: ' + data.host + '\n';
+    res += 'chip id: ' + data.cid + '\n';
+    res += 'node id: ' + data.nid + '\n';
+    res += 'replica id: ' + data.rid + '\n';
+    if (parent.selectedMetricIdx >= 0) {
+      res += parent.metrics[parent.selectedMetricIdx].label + ' spends ' +
+          data.value.toFixed(2) + 'us in total, ';
+      const pcnt = data.value / data.total * 100;
+      res += 'taking ' + pcnt.toFixed(2) + '% of a step.';
+    }
+    return res;
+  },
+  /**
+   * Draw the legend of the graph.
+   */
+  drawLegend: function(
+      svg: any, height: number, legendElementHeight: number,
```

Comment:
`svg: any` seems off. Perhaps you meant `svg: SVGElement`?

Reply:
Change to SVGElement does not build. Raises this error "Property 'selectAll' does not exist on type 'SVGElement'."